src/lib/visualization/join_graph_visualizer.cpp

```diff
@@ -0,0 +1,126 @@
+#include "join_graph_visualizer.hpp"
+
+#include <sstream>
+
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+#include "viz_record_layout.hpp"
+
+namespace opossum {
+
+void JoinGraphVisualizer::_build_graph(const std::vector<JoinGraph>& graphs) {
+  for (const auto& graph : graphs) {
+    for (auto vertex_idx = size_t{0}; vertex_idx < graph.vertices.size(); ++vertex_idx) {
+      const auto& vertex = graph.vertices[vertex_idx];
+      const auto predicates = graph.find_local_predicates(vertex_idx);
+
+      VizRecordLayout layout;
+      layout.add_label(_create_vertex_description(vertex));
+
+      if (!predicates.empty()) {
+        auto& predicates_layout = layout.add_sublayout();
+
+        for (const auto& predicate : predicates) {
+          predicates_layout.add_label(predicate->as_column_name());
+        }
+      }
+
+      VizVertexInfo vertex_info = _default_vertex;
+      vertex_info.label = layout.to_label_string();
+      vertex_info.shape = "record";
+
+      // Don't wrap the label, we're using the record layout, which has internal formatting. Randomly inserted newlines
+      // wouldn't help here
+      _add_vertex(vertex, vertex_info, WrapLabel::Off);
+    }
+
+    for (const auto& edge : graph.edges) {
+      const auto vertex_count = edge.vertex_set.count();
+
+      // Single-vertex edges are local predicates. We already visualized those above
+      if (vertex_count <= 1) continue;
+
+      // Binary vertex edges: render a simple edge between the two vertices with the predicates as the edge label
+      if (vertex_count == 2) {
+        const auto first_vertex_idx = edge.vertex_set.find_first();
+        const auto second_vertex_idx = edge.vertex_set.find_next(first_vertex_idx);
+
+        const auto first_vertex = graph.vertices[first_vertex_idx];
+        const auto second_vertex = graph.vertices[second_vertex_idx];
+
+        std::stringstream edge_label_stream;
+        for (const auto& predicate : edge.predicates) {
+          edge_label_stream << predicate->as_column_name();
+          edge_label_stream << "\n";
+        }
+
+        VizEdgeInfo edge_info;
+        edge_info.color = _random_color();
+        edge_info.font_color = edge_info.color;
+        edge_info.dir = "none";
+        edge_info.label = edge_label_stream.str();
+
+        _add_edge(first_vertex, second_vertex, edge_info);
+      } else {
+        // More than two vertices, i.e. we have a hyperedge. Render a diamond vertex that contains all the Predicates
```

Comment:
Could you give an example of an SQL query that produces such a hyperedge?

Reply:
Added `SELECT * FROM x, y, z WHERE x.a + y.b + z.c = x.d`